### PR TITLE
feat: find method for DocumentArray

### PR DESCRIPTION
**Pull request by davidbp**
Allow users to find Documents in a  DocumentArray that match several regular expressions per tag

### CODE REVIEW

File: jina/types/arrays/search_ops.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            regexes[tag_name] = re.compile(regex)
+
+        for pos, doc in enumerate(iterdocs):
+            for tag_name, pattern in regexes.items():
+                tag_value = doc.tags.get(tag_name, None)
+                if tag_value:
+                    if pattern.match(tag_value):
+                        matched_counts[pos] += 1
+
+        if operator == '<':
```
Comment (JoanFM) [inline jina/types/arrays/search_ops.py:58]: keep this in a dictionary mapping from `operator_name` to operation function or partialfunc

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/types/arrays/search_ops.py (hunk 1 above)
Comment (davidbp) [reply jina/types/arrays/search_ops.py]: good catch, done

File: jina/types/arrays/search_ops.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        '>=': operator.ge,
+    }
+
+    def find(
+        self,
+        regexes: Dict,
+        traversal_paths: list = ['r'],
+        operator: str = '>=',
+        value: int = 1,
+    ) -> None:
```
Comment (tadejsv) [inline jina/types/arrays/search_ops.py:26]: You are not returning `None`

File: jina/types/arrays/search_ops.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        """
+        Find Documents that match the regular expressions in `regexes`.
```
Comment (tadejsv) [inline jina/types/arrays/search_ops.py:28]: You might want to reword this to `Find Documents whose tags match` - on first reading it was not clear to me what these regexes are compared to.

File: jina/types/arrays/search_ops.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        for pos, doc in enumerate(iterdocs):
+            for tag_name, pattern in regexes.items():
+                tag_value = doc.tags.get(tag_name, None)
+                if tag_value:
+                    if pattern.match(tag_value):
+                        matched_counts[pos] += 1
+
+        operator_func = self.operators[operator]
+        coordinate_flags = operator_func(matched_counts, value)
+        indices = np.where(coordinate_flags)[0].tolist()
```
Comment (tadejsv) [inline jina/types/arrays/search_ops.py:64]: Maybe that's just a matter of taste, but it seems to me that it would be simpler to add documents to filtered list in the "main" loop - seems like gains from vectorization here would be quite small.

File: jina/types/arrays/search_ops.py (hunk 3 above)
Comment (davidbp) [reply jina/types/arrays/search_ops.py:28]: I totally agree

File: jina/types/arrays/search_ops.py (hunk 4 above)
Comment (davidbp) [reply jina/types/arrays/search_ops.py]: Agree, the regular expression match takes longer than anything else.

File: tests/unit/types/arrays/test_search_ops.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    d2 = Document(tags={'city': 'Berlin', 'phone': '648907348'})
+    d3 = Document(tags={'city': 'Paris', 'phone': 'None'})
+    d4 = Document(tags={'city': 'Brussels', 'phone': 'None'})
+    docs = [d1, d2, d3, d4]
+    return docs
+
+
+def test_single_regex(list_doc_examples, tmpdir):
+    regexes = {'city': r'B.*'}
+    D = DocumentArray(list_doc_examples)
```
Comment (JoanFM) [inline tests/unit/types/arrays/test_search_ops.py:20]: if possible, do not use capital letters as variables

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/types/arrays/test_search_ops.py (hunk 5 above)
Comment (davidbp) [reply tests/unit/types/arrays/test_search_ops.py:20]: I will do it but I don't feel it's a great idea, capitals for Matrices/Tensors/DocumentArrays make a lot of sense, non capital versions for vectors/documents also make sense. Makes everything easier to read.

Comment (JoanFM) [reply tests/unit/types/arrays/test_search_ops.py:20]: this is standard programming standards, capitals are reserved for class names and not for vars or instances

File: jina/types/arrays/search_ops.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        '>': operator.gt,
+        '==': operator.eq,
+        '!=': operator.ne,
+        '<=': operator.le,
+        '>=': operator.ge,
+    }
+
+    def find(
+        self,
+        regexes: Dict,
```
Comment (JoanFM) [inline jina/types/arrays/search_ops.py:24]: are we sure this complies with the definition of a `regex`?

File: jina/types/arrays/search_ops.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        traversal_paths: list = ['r'],
+        operator: str = '>=',
```
Comment (JoanFM) [inline jina/types/arrays/search_ops.py:26]: I do not understand what this operator is for

File: tests/unit/types/arrays/test_search_ops.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    filtered_doc_array = doc_array.find(regexes=regexes)
+    filtered_doc_array_memmap = doc_array_memmap.find(regexes=regexes)
+
+    # Examples with Barcelona, Berlin, Brussels should match
+    assert len(filtered_doc_array) == 3
+    assert len(filtered_doc_array_memmap) == 3
+
+
+def test_multiple_regex(list_doc_examples, tmpdir):
```
Comment (JoanFM) [inline tests/unit/types/arrays/test_search_ops.py:32]: instead of doing the test twice, please parametrize with DocArray and DocArrayMemmap

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/types/arrays/search_ops.py (hunk 6 above)
Comment (davidbp) [reply jina/types/arrays/search_ops.py:24]: I guess you mean `Dict[str,str]` ?

File: jina/types/arrays/search_ops.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        value: int = 1,
```
Comment (JoanFM) [inline jina/types/arrays/search_ops.py:27]:
```suggestion
        value: Optional[int] = None,
```

And then in the code say.

```python
value = value or len(regexes)
```

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/types/arrays/test_search_ops.py (hunk 8 above)
Comment (davidbp) [reply tests/unit/types/arrays/test_search_ops.py]: Done

File: jina/types/arrays/search_ops.py (hunk 9 above)
Comment (davidbp) [reply jina/types/arrays/search_ops.py]: yes, done

File: jina/types/arrays/search_ops.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+            operator in self.operators
+        ), f'operator={operator} is not a valid operator from {self.operators.keys()}'
+
+        operator_func = self.operators[operator]
+        iterdocs = self.traverse_flat(traversal_paths)
+        filtered = DocumentArray()
+
+        value = value or len(regexes)
+
+        for tag_name, regex in regexes.items():
```
Comment (JoanFM) [inline jina/types/arrays/search_ops.py:61]: I wonder if the interface should be this or provide already `compiled` regexes? Can we check if the items are already `regexes` and we do not need to `compile` them? This way we could actually use in an executor

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/types/arrays/search_ops.py (hunk 10 above)
Comment (davidbp) [reply jina/types/arrays/search_ops.py:61]: I think it is more convenient for the user to simply pass a string rather than having to compile them

File: tests/unit/types/arrays/test_search_ops.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+
+@pytest.mark.parametrize('doc_array_creator', [docarray_type, docarray_memmap_type])
+def test_single_regex(doc_array_creator, list_doc_examples, tmpdir):
+    regexes = {'city': r'B.*'}
+    doc_array = doc_array_creator(list_doc_examples, tmpdir)
+    filtered_doc_array = doc_array.find(regexes=regexes)
+
+    # Examples with Barcelona, Berlin, Brussels should match
+    assert len(filtered_doc_array) == 3
```
Comment (bwanglzu) [inline tests/unit/types/arrays/test_search_ops.py:35]: can you assert the result is correct?

File: tests/unit/types/arrays/test_search_ops.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+def test_multiple_regex(doc_array_creator, list_doc_examples, tmpdir):
+    regexes = {'city': r'B.*', 'phone': 'Non'}
+    doc_array = doc_array_creator(list_doc_examples, tmpdir)
+
+    filtered_doc_array = doc_array.find(
+        regexes=regexes, traversal_paths=['r'], operator='==', value=2
+    )
+
+    # Examples with Barcelona, Brussels should match
+    assert len(filtered_doc_array) == 2
```
Comment (bwanglzu) [inline tests/unit/types/arrays/test_search_ops.py:48]: same as above

**Review by bwanglzu — CHANGES_REQUESTED** [review]

File: jina/types/arrays/search_ops.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            if type(regex) == str:
```
Comment (JoanFM) [inline jina/types/arrays/search_ops.py:62]:
```suggestion
            isinstance(regex, str):
```

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: .github/2.0/cookbooks/Document.md [hunk 14]
```diff
@@ -1001,6 +1001,62 @@ match emb = [1.  0.1 0.  0.  0. ] score = 1.6763054132461548
...
+d1 = Document(tags={'city': 'Barcelona', 'phone':'None'})
+d2 = Document(tags={'city': 'Berlin','phone':'648907348'})
+d3 = Document(tags={'city': 'Paris', 'phone': 'None'})
+d4 = Document(tags={'city': 'Brussels', 'phone': 'None'})
+
+docarray = DocumentArray([d1, d2, d3, d4])
+
+regexes = {'city':r'B.*'}
+docarray_filtered = docarray.find(regexes=regexes)
+print(f'len(docarray_filtered)={len(docarray_filtered)}')
```
Comment (JoanFM) [inline .github/2.0/cookbooks/Document.md:1022]: see how in the rest of the cookbook we show output without actually writing `print`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: .github/2.0/cookbooks/Document.md (hunk 14 above)
Comment (davidbp) [reply .github/2.0/cookbooks/Document.md:1022]: I see it full of prints! Besides, as  a reader isn't it better to get pretty printed results?

File: tests/unit/types/arrays/test_search_ops.py (hunk 11 above)
Comment (davidbp) [reply tests/unit/types/arrays/test_search_ops.py:35]: yes, done

**Review by bwanglzu — APPROVED** [review]
LGTM!

**Review by JoanFM — APPROVED** [review]
test with regex as values in the `regex` dict
